<html>
<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
</head>
<body lang=EN-GB link=blue vlink=purple>
<h2>Software and Scripts.</h2>

<h4>Ian Wilson</h4>


Software and scripts used in the paper paper <b>Titin founder mutation is a common
  cause of myofibrillar myopathy with early respiratory failure</b>.  2014.  Pfeffer <em> et al.</em>
 <a href="http://jnnp.bmj.com/content/85/3/331.short">link</a>.  These have all been tested on ubuntu and red hat linux.  .  The tools tabix, vcftools and R are available as part of the ubuntu package management system and
can be installed using <code>sudo apt-get install tabix vcftools rbase </code>

<ul>
<li><a href="http://samtools.sourceforge.net/">tabix</a> available as part of the <a href="http://sourceforge.net/projects/samtools/">samtools</a> package.</li>
<li><a href="http://vcftools.sourceforge.net/">vcftools</a></li>
<li><a href="http://stephenslab.uchicago.edu/software.html">PHASE</a>
<li><a href="http://www.broad.mit.edu/mpg/haploview/">haploview</a></li>
<li>The <a href="www.r-project.org" >R</a> statistical programming environment.</li>
</ul>


The custom written scripts and R code used are
<ul>
<li>
<a href="1000_genomes_script" >1000_genomes_script</a>.  Uses tabix and vcftools to download.
<li>
<a href="filter1000genomes.py" >filter1000genomes.py</a>
<li>
<a href="Transform_PHASE.R" >Transform_PHASE.R</a>
<li>
<a href="haploview_phase.R" >haploview_phase.R</a>
<li>
<a href="readphase.R">readphase.R</a></li> Read the output of phase pairs into R.
</ul>

Data Files
<ul>
<li><a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/supporting/phase1_samples_integrated_20101123.ped">
phase1_samples_integrated_20101123.ped
</a>
downloaded from the 1000 genomes data repository.
</ul>


<p>The 100 genomes data was <span style='color:windowtext'>downloaded
between positions 178404570 and 179734924 using <b>tabix</b>.  This was then
converted to the IMPUTE format for phased data by <b>vcftools</b> (<a
href="http://vcftools.sourceforge.net/">vcftools.sourceforge.net</a>).
Filtering this data by only including positions shared with our assays was performed using a custom written <b>python</b> script,
 <a href="filter1000genomes.py" >filter1000genomes.py</a>, and
then the data was prepared for the programs <a
href="http://stephenslab.uchicago.edu/software.html">PHASE</a> and  <a
href="http://www.broad.mit.edu/mpg/haploview/">haploview/</a> using <b>R</b>
scripts, selecting only those individuals from the CEU and GB populations
that did not have parents in the data set, and by adding homozygote wild
type alleles at the disease position
 for 1000 genomes data.</p>

<p>
  Sample
 information was obtained from the file
 <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/supporting/phase1_samples_integrated_20101123.ped">
 phase1_samples_integrated_20101123.ped</span></a><span
 style='color:windowtext'> downloaded from the 1000 genomes data repository.
</p>

<b>PHASE</b> was used to phase haplotypes, using the known
haplotypes option to use the 1000 genomes data as known phase data. The pairs
output file was used to determine the relative posterior probability of the
different haplotype reconstructions for these data. <b>R</b> was used for post
processing of <b>PHASE</b> pairs output. The <b>haploview</b> program was used
to investigate the haplotype structure of the population data and to produce
plots.
</p>
<hr>
<h3>Workflow</h3>

Once you have all the tools installed, then the analysis done in the paper can
be repeated with the following commands.

<ul>
  <li>
    Create a working directory and copy the files <code>1000_genomes_script</code>,
    <code>haploview_phase.R</code>, <code>Transform_PHASE.R</code>,
    <code>HMERFmatch1K.csv</code>,  <code>HMERF_Haplotype.csv </code>,
    <code>filter1000genomes.py</code> and <code>readphase.R</code>.
  </li>
  <li>
    Run the <code>1000_genomes_script</code> using <br />
    <code>bash 1000_genomes_script</code>
  </li>
  <li>
    <code>R --vanilla readphase.R</code>
  </li>
</ul>
<hr />
<p>
  <a href="mailto:ian.wilson@newcastle.ac.uk">ian.wilson@newcastle.ac.uk</a>
</p>
</body>
</html>
