---
title: "Shared Haplotypes around a mutation"
author: "Ian Wilson"
date: "30 January 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### obtain the pedigree information for the 1000 genomes samples

```
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/supporting/phase1_samples_integrated_20101123.ped
```

## use tabix to get a copy of the data between positions 178404570 and 179734924 on chromosome 2

```
tabix -fh ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/integrated_call_sets/ALL.chr2.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.vcf.gz 2:178404570-179734924 > temporary.vcf
```

#### transform to get phased output in IMPUTE format

```
vcftools --vcf temporary.vcf --IMPUTE --out impute1K 
```

Examine these positions a bit more

```{r read_positions}
positions_1K <- read.table("impute1K.impute.legend", header=TRUE)
positions_HMERF <- read.csv("HMERF_Haplotype.csv")
```
we have read `r nrow(positions)` positions from the 1000 genomes.


```{python}
import sys
data=[a.strip().split(",") for a in open("HMERF_Haplotype.csv")]

header, data = data[0], data[1:]
outHMERF = open("HMERFmatch1K.csv","w")
outHMERF.write(",".join(h for h in header)+"\n")
## step through the positions available 
positions=dict()

for index,b in enumerate(data):
    positions[int(b[0])] = (index,b)

infile = open("impute1K.impute.legend")
header=infile.readline()
outlegend = open("match1K.legend","w")
outlegend.write(header)

seen=set()
indices=dict()
print "=============================\nSNPs present in both"
print "-----------------------------"
for linenum,a in enumerate(infile):
    ID,pos,allele0,allele1 = a.strip().split()
    if int(pos) in positions:
        outlegend.write(a)
        print ID,pos,allele0,allele1,"| HMERF: "," ".join(v for v in positions[int(pos)][1])
        seen.add(int(pos))
        indices[linenum] = ([allele0,allele1],positions[int(pos)][0])
   

infile.close()
outlegend.close()

missing=[]
print "=============================\nMissing in 1K"
print "-----------------------------"

for k,v in positions.iteritems():
    if k not in seen:
        missing.append(positions[k])
        print " ".join(b for b in v[1])

seen.add(179410829)
pos = [s for s in seen]
pos.sort()
print pos

for p in pos:
    if p in seen:
        outHMERF.write(",".join(v for v in positions[p][1])+"\n")

outHMERF.close()

missing=sorted(missing)
print "============================="
# now lets try to filter the data
infile = open("impute1K.impute.hap")
outfile = open("match1K","w")
for index,line in enumerate(infile):
    if index in indices:
        b = line.strip().split()
        code =   indices[index][0]
        outfile.write(" ".join(code[int(a)] for a in b)+"\n")

outfile.close()
infile.close()
```



### Prepare data for PHASE


```{r}
source("Transform_PHASE.R")
```
We now have 


